keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1255 @@
+
+func (etv *enumTagValidator) GetValidations(context Context, _ []string, payload string) (Validations, error) {
+        if context.Type != types.String {
+                return Validations{}, fmt.Errorf("can only be used on string types")
+        }
+
+        var result Validations
+
+        if enum, ok := etv.enumContext.EnumType(context.Parent); ok {
+                supportVarName := PrivateVar{Name: "SymbolsFor" + context.Parent.Name.Name, Package: "local"}
+                supportVar := Variable(supportVarName, GenericFunction(enumTagName, DefaultFlags, setsNew, []types.Name{enum.Name}, enum.ValueArgs()...))
+                result.AddVariable(supportVar)
+                fn := Function("enum", DefaultFlags, enumValidator, supportVarName)
+                result.AddFunction(fn)
+        }
+
+        return result, nil
+}
+
+func init() {
+    AddToRegistry(EnumValidator) // Registers +enum
+}
+```
+
+validation-gen processes types.go files, searching for tags. For example:
+
+```go
+type DeploymentStrategy struct {
+    Type DeploymentStrategyType
+    …
+}
+// +k8s:enum // <---- THIS IS A validation-gen TAG
+type DeploymentStrategyType string
+const (
+    RecreateDeploymentStrategyType DeploymentStrategyType = "Recreate"
+    RollingUpdateDeploymentStrategyType DeploymentStrategyType = "RollingUpdate"
+)
+```
+
+When a tag is found by validation-gen.  validation-gen uses the plugins to generate the appropriate code:
+
+```go
+// generated
+var symbolsForDeploymentStrategy = sets.New[E1](v1.Recreate, v1.RollingUpdate)
+
+func Validate_DeploymentStrategy(in *v1.DeploymentStrategy, 
+        fldPath *field.Path) (errs field.ErrorList) {
+    errs = append(errs, validation.ValidateEnum(
+        fldPath.Child("type"), in.Type, symbolsForDeploymentStrategy)...)
+    return errs
+}
+```
+
+The generator will also auto register all validations in the runtime.Scheme.
+
+Once validation is generated, it will be easy to opt-in (see below snippet).
+
+**NOTE:** This does not actually do anything until the tags are used on types/fields, and all tags are net-new. None of the existing tags cause any validation code to be generated as designed by having our own bespoke tags (eg: `+k8s:required`(new w/ `validation-gen`) vs `+required`(old), etc.)
+
+```go
+// +k8s:validation-gen=TypeMeta
+// +k8s:validation-gen-input=k8s.io/api/apps/v1
+```
+
+### Catalog of Supported Validation Rules & Associated IDL Tags
+A number of the rules in the below sections are not implemented but will be trivial to implement once we are are aligned on the right pattern/syntax for the given validator. Implementing a validator “just-in-time” means we don't add more without real fields using them. We estimate that in the limit we may have 30-40 validators, but today we have less than 10.  
+
+The below rules are currently implemented or are very similar to an existing validator in the [validation-gen prototype](prototype )
```

Comment:
This link (= https://github.com/kubernetes/enhancements/blob/abd9ff7fa280923164f2e8d049771f9cabab7be4/keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/prototype) is invalid.

I was looking for more information on union value support because I have a specific use case and question regarding them over in https://github.com/kubernetes/enhancements/pull/5034/files#r1935225480. @aaron-prindle, I tagged you there.

Reply:
Apologies, thanks for flagging this. This is fixed in the KEP now.

The current prototype is is jpbetz/kubernetes in the `validation-gen` feature branch.  Link here:
https://github.com/jpbetz/kubernetes/tree/validation-gen